src/order/directed.lean

```diff
@@ -47,21 +48,44 @@ theorem directed.mono_comp {ι} {rb : β → β → Prop} {g : α → β} {f : 
   directed rb (g ∘ f) :=
 directed_comp.2 $ hf.mono hg
 
-/-- A monotone function on a sup-semilattice is directed. -/
-lemma directed_of_sup [semilattice_sup α] {f : α → β} {r : β → β → Prop}
+/-- A `preorder` is a `directed_order` if for any two elements `i`, `j`
+there is an element `k` such that `i ≤ k` and `j ≤ k`. -/
+class directed_order (α : Type u) extends preorder α :=
+(directed : ∀ i j : α, ∃ k, i ≤ k ∧ j ≤ k)
+
+@[priority 100]  -- see Note [lower instance priority]
+instance semilattice_sup.to_directed_order (α) [semilattice_sup α] : directed_order α :=
+⟨λ i j, ⟨i ⊔ j, le_sup_left, le_sup_right⟩⟩
+
+/-- A monotone function on a directed order (aka a sup-semilattice) is directed. -/
```

Comment:
I think this description is a little misleading in that it says directed orders are also known as sup-semilattices: perhaps better phrasing would be "on a directed order (in particular a sup-semilattice) is".

Reply:
I guess I was trying to hand-wave over the asymmetry with the `_inf` version below. I'm not really convinced `directed_order` is useful, since it's really just `semilattice_sup` with a non-computable `sup`

Reply:
Directed orders are very useful! And they're very different from just semilattice_sup: in particular in a directed order there's no guarantee that the given element is a supremum of the two given elements (ie `sup_le` might not hold), see eg [here](https://en.wikipedia.org/wiki/Directed_set#Contrast_with_semilattices).

Reply:
Well, I'd argue they can't be that useful because mathlib has no instances of them that aren't linear orders so far!

Reply:
But thanks for clearing that up, I was missing the requirement for uniqueness.